### PR TITLE
Add support for marshalling lists of strings in HTTP headers...

### PR DESCRIPTION
**Pull request by Bennett-Lynch**
...for JSON/XML protocols.

## Motivation and Context

We currently lack support for marshalling lists of strings (and enums)
in HTTP headers. Other IDLs and modeling languages, e.g., Smithy, do
support such bindings and services may come to expect such support:

https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpheader-trait

## Description

* Add support for marshalling a list of strings to relevant JSON and XML
marshalling classes
* Consistent with the existing XML HeaderUnmarshaller map
implementation, unmarshalling only supports string value types (this can
be extended in the future if needed)
* Marshalling implementations are changed to use appendHeader rather
than putHeader
* Add convenience method SdkField#getRequiredTrait(..) to support the
common use case of ensuring a trait is present; update other existing
use cases to utilize this method
* Update relevant test classes to correctly recognize headers as a
`Map<String, List<String>>`
* Only wrap marshalling assertion exceptions as runtime when needed,
minimizing stack trace noise

## Testing

New tests added to:
* rest-json-input.json
* rest-json-output.json
* rest-xml-input.json
* rest-xml-output.json

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/HeaderMarshaller.java [hunk 1]
```diff
@@ -66,6 +68,24 @@ protected boolean shouldEmit(Map map) {
...
+            SdkField memberFieldInfo = sdkField.getRequiredTrait(ListTrait.class).memberFieldInfo();
+            for (Object listValue : list) {
+                XmlMarshaller marshaller = context.marshallerRegistry().getMarshaller(MarshallLocation.HEADER, listValue);
+                marshaller.marshall(listValue, context, paramName, memberFieldInfo);
+            }
+        }
+
+        @Override
+        protected boolean shouldEmit(List list) {
+            return list != null && !list.isEmpty();
```
Comment (zoewangg) [inline core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/HeaderMarshaller.java:86]: nit: should we just use CollectionUtils.isNullOrEmpty?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/HeaderMarshaller.java [hunk 2]
```diff
@@ -45,6 +48,17 @@
     public static final JsonMarshaller<Instant> INSTANT
         = new SimpleHeaderMarshaller<>(JsonProtocolMarshaller.INSTANT_VALUE_TO_STRING);
 
+    public static final JsonMarshaller<List<?>> LIST = (list, context, paramName, sdkField) -> {
+        if (list.isEmpty()) {
```
Comment (zoewangg) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/HeaderMarshaller.java:52]: Do we need to check null here?

**Review by zoewangg — COMMENTED** [review]
Do we need to support it in query protocol?

**Comment by Bennett-Lynch** [comment]
> Do we need to support it in query protocol?

Discussed offline and we believe this is not applicable to query protocol.

**Review by zoewangg — APPROVED** [review]

**Review by joviegas — APPROVED** [review]

**Pull request merged**